client/client.go

```diff
@@ -283,138 +287,116 @@ func (c *client) tsLoop() {
+				go func(dc string, tsoDispatcher chan *tsoRequest) {
+					var (
+						err       error
+						stream    pdpb.PD_TsoClient
+						opts      []opentracing.StartSpanOption
+						requests  = make([]*tsoRequest, maxMergeTSORequests+1)
+						createdCh = make(chan struct{})
+					)
+					for {
+						ctx, cancel := context.WithCancel(loopCtx)
+						if stream == nil {
+							go c.checkStreamTimeout(loopCtx, cancel, createdCh)
+							stream, err = pdpb.NewPDClient(c.connMu.clientConns[c.connMu.allocators[dc]]).Tso(ctx)
+							if stream != nil {
+								createdCh <- struct{}{}
+							}
+							if err != nil {
+								select {
+								case <-loopCtx.Done():
+									cancel()
+									return
+								default:
+								}
+								log.Error("[pd] create tso stream error", zap.String("dc-location", dc), errs.ZapError(errs.ErrClientCreateTSOStream, err))
+								c.ScheduleCheckLeader()
+								cancel()
+								c.revokeTSORequest(errors.WithStack(err), tsoDispatcher)
+								select {
+								case <-time.After(time.Second):
+								case <-loopCtx.Done():
+									cancel()
+									return
+								}
+								continue
+							}
+						}
+						select {
+						case first := <-tsoDispatcher:
+							pendingPlus1 := len(tsoDispatcher) + 1
+							requests[0] = first
+							for i := 1; i < pendingPlus1; i++ {
+								requests[i] = <-tsoDispatcher
+							}
+							done := make(chan struct{})
+							dl := deadline{
+								timer:  time.After(c.timeout),
+								done:   done,
+								cancel: cancel,
+							}
+							c.tsDeadline.RLock()
+							tsDeadlineCh := c.tsDeadline.ch[dc]
+							c.tsDeadline.RUnlock()
+							select {
+							case tsDeadlineCh <- dl:
+							case <-loopCtx.Done():
+								cancel()
+								return
+							}
+							opts = extractSpanReference(requests[:pendingPlus1], opts[:0])
+							err = c.processTSORequests(stream, dc, requests[:pendingPlus1], opts)
+							close(done)
+						case <-loopCtx.Done():
+							cancel()
+							return
+						}
+						if err != nil {
+							select {
+							case <-loopCtx.Done():
+								cancel()
+								return
+							default:
+							}
+							log.Error("[pd] getTS error", zap.String("dc-location", dc), errs.ZapError(errs.ErrClientGetTSO, err))
+							c.ScheduleCheckLeader()
+							cancel()
+							stream = nil
+						}
+					}
+				}(dcLocation, c.tsoDispatcher.ch[dcLocation])
 			}
-			log.Error("[pd] getTS error", errs.ZapError(errs.ErrClientGetTSO, err.Load().(error)))
-			c.ScheduleCheckLeader()
-			cancel()
-			streams, cancel = make(map[string]pdpb.PD_TsoClient), nil
 		}
 	}
 }
 
-func (c *client) checkTSOStreams(streams map[string]pdpb.PD_TsoClient) bool {
-	dcLocations := c.getDCLocations()
-	if len(streams) != len(dcLocations) {
+func (c *client) checkTSODispatcher(dcLocation string) bool {
+	c.tsoDispatcher.RLock()
+	defer c.tsoDispatcher.RUnlock()
+	tsoChan, exist := c.tsoDispatcher.ch[dcLocation]
+	if !exist || tsoChan == nil {
```

Comment:
When will the tsoChan become nil?

Reply:
According to the existing code, this will not happen.